packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -1440,46 +1444,102 @@ class BuildOwner {
+  /// calls the given `callback`.
+  ///
+  /// This mechanism is used to ensure that, for instance, [State.dispose] does
+  /// not call [State.setState].
+  void lockState(void callback()) {
+    assert(callback != null);
+    assert(_debugStateLockLevel >= 0);
+    assert(() {
+      _debugStateLockLevel += 1;
+      return true;
+    });
+    try {
+      callback();
+    } finally {
+      assert(() {
+        _debugStateLockLevel -= 1;
+        return true;
+      });
+    }
+    assert(_debugStateLockLevel >= 0);
+  }
+
+  /// Establishes a scope for updating the widget tree, and calls the given
+  /// `callback`, if any. Then, builds all the elements that were marked as
+  /// dirty using [scheduleBuildFor], in depth order.
   ///
   /// This mechanism prevents build functions from transitively requiring other
   /// build functions to run, potentially causing infinite loops.
   ///
-  /// If the building argument is true, then this function enables additional
-  /// asserts that check invariants that should apply during building.
+  /// The dirty list is processed after `callback` returns, building all the
+  /// elements that were marked as dirty using [scheduleBuildFor], in depth
+  /// order. If elements are marked as dirty while this method is running, they
+  /// must be deeper than the `context` node, and deeper than any
+  /// previously-built node in this pass.
+  ///
+  /// To flush the current dirty list without performing any other work, this
+  /// function can be called with no callback. This is what the framework does
+  /// each frame, in [WidgetsBinding.beginFrame].
   ///
-  /// The context argument is used to describe the scope in case an exception is
-  /// caught while invoking the callback.
-  void lockState(void callback(), { bool building: false }) {
-    bool debugPreviouslyBuilding;
+  /// Only one [buildScope] can be active at a time.
+  ///
+  /// A [buildScope] implies a [lockState] scope as well.
+  void buildScope(Element context, [VoidCallback callback]) {
+    if (callback == null && _dirtyElements.isEmpty)
+      return;
+    assert(context != null);
     assert(_debugStateLockLevel >= 0);
+    assert(!_debugBuilding);
     assert(() {
-      if (building) {
-        debugPreviouslyBuilding = _debugBuilding;
-        _debugBuilding = true;
-      }
       _debugStateLockLevel += 1;
+      _debugBuilding = true;
       return true;
     });
+    Timeline.startSync('Build');
     try {
-      callback();
+      _scheduledFlushDirtyElements = true;
+      if (callback != null)
+        callback();
+      _dirtyElements.sort(_elementSort);
+      int dirtyCount = _dirtyElements.length;
+      int index = 0;
+      while (index < dirtyCount) {
+        assert(_dirtyElements[index] != null);
+        assert(!_dirtyElements[index]._active || _dirtyElements[index]._debugIsInScope(context));
+        _dirtyElements[index].rebuild();
+        index += 1;
+        if (dirtyCount < _dirtyElements.length) {
```

Comment:
well that's a frightening thought, but I guess you're right, it could. If everything in the tree was marked dirty (e.g. because some inherited guy changed), and then some global key shenanigans caused a node to be moved around, we could end up doing them in the wrong order. I'll file a separate bug to track that. https://github.com/flutter/flutter/issues/5757